File: app/src/shared/helper/observability.ts
import { z } from 'zod';
import {
  HelperKeyboardOwnerSnapshotSchema,
  HelperDiagnosticIdentitySchema,
} from './protocol-values';

const AggregateCounterSchema = z.number().int().min(0).max(Number.MAX_SAFE_INTEGER);
const EffectOutcomeCountersSchema = z
  .object({
    attempted: AggregateCounterSchema,
    succeeded: AggregateCounterSchema,
    partial: AggregateCounterSchema,
    failed: AggregateCounterSchema,
  })
  .strict();
export const HelperOwnerObservabilitySchema = z
  .object({
    starts: AggregateCounterSchema,
    cleanExits: AggregateCounterSchema,
    abnormalExits: AggregateCounterSchema,
    singletonCollisions: AggregateCounterSchema,
    authAttempts: AggregateCounterSchema,
    authFailures: z
      .object({
        crossUser: AggregateCounterSchema,
        wrongSession: AggregateCounterSchema,
        codeIdentity: AggregateCounterSchema,
        mac: AggregateCounterSchema,
        protocol: AggregateCounterSchema,
      })
      .strict(),
    leaseAcquired: AggregateCounterSchema,
    leaseRenewed: AggregateCounterSchema,
    leaseExpired: AggregateCounterSchema,
    leaseDisconnected: AggregateCounterSchema,
    leaseReleasedNeutral: AggregateCounterSchema,
    leaseReleasedDraining: AggregateCounterSchema,
    drainDurationMsTotal: AggregateCounterSchema,
    drainDurationMsMax: AggregateCounterSchema,
    maintenancePostponed: AggregateCounterSchema,
    handoffSucceeded: AggregateCounterSchema,
    handoffFailed: AggregateCounterSchema,
    degraded: AggregateCounterSchema,
    hookRecoveries: AggregateCounterSchema,
  })
  .strict()
  .refine((owner) => owner.drainDurationMsMax <= owner.drainDurationMsTotal, {
    message: 'Maximum owner drain duration cannot exceed total duration',
    path: ['drainDurationMsMax'],
  });
export const HelperRuntimeObservabilitySchema = z
  .object({
    keyboardOwner: HelperKeyboardOwnerSnapshotSchema,
    owner: HelperOwnerObservabilitySchema,
    registeredInput: z
      .object({
        hookInstalled: AggregateCounterSchema,
        pumpAlive: AggregateCounterSchema,
        hcActionCallbacks: AggregateCounterSchema,
        physicalCallbacks: AggregateCounterSchema,
        physicalCallbacksFiltered: AggregateCounterSchema,
        registeredCandidateCallbacks: AggregateCounterSchema,
        registeredMatchCallbacks: AggregateCounterSchema,
        registeredReleaseCallbacks: AggregateCounterSchema,
        callbackChannelAccepted: AggregateCounterSchema,
        callbackChannelRejected: AggregateCounterSchema,
        adapterDequeued: AggregateCounterSchema,
        ownerAdmitted: AggregateCounterSchema,
        ownerFlushed: AggregateCounterSchema,
        ownerRejected: AggregateCounterSchema,
        gatewayReceived: AggregateCounterSchema,
        v10NotificationAccepted: AggregateCounterSchema,
        electronReceived: AggregateCounterSchema,
        observationAccepted: AggregateCounterSchema,
      })
      .strict(),
    keyboardCapture: z
      .object({
        runtimeRollbackActive: z.boolean(),
        developmentDisabled: z.boolean(),
        activationEnableRequestsBlocked: AggregateCounterSchema,
        sessionCaptureRequestsBlocked: AggregateCounterSchema,
        shutdownOwnershipDeadlines: AggregateCounterSchema,
        terminalDisablements: AggregateCounterSchema,
      })
      .strict(),
    transactions: z
      .object({
        started: AggregateCounterSchema,
        committed: AggregateCounterSchema,
        replayed: AggregateCounterSchema,
        cancelled: AggregateCounterSchema,
        journalHighWater: AggregateCounterSchema,
        cancellationReasons: z
          .object({
            invalidContinuation: AggregateCounterSchema,
            modifierChanged: AggregateCounterSchema,
            altGr: AggregateCounterSchema,
            journalOverflow: AggregateCounterSchema,
            configurationReplaced: AggregateCounterSchema,
            revisionMismatch: AggregateCounterSchema,
            gateClosed: AggregateCounterSchema,
            shutdown: AggregateCounterSchema,
            helperDisconnected: AggregateCounterSchema,
            secureDesktop: AggregateCounterSchema,
            timeout: AggregateCounterSchema,
            activationDeliveryFailed: AggregateCounterSchema,
            neutralizationFailed: AggregateCounterSchema,
            replayFailed: AggregateCounterSchema,
            effectProtocolViolation: AggregateCounterSchema,
            physicalStateMismatch: AggregateCounterSchema,
            targetChanged: AggregateCounterSchema,
          })
          .strict(),
      })
      .strict(),
    replay: EffectOutcomeCountersSchema,
    dummy: EffectOutcomeCountersSchema,
    paste: z
      .object({
        attempted: AggregateCounterSchema,
        submitted: AggregateCounterSchema,
        targetValidationFallback: AggregateCounterSchema,
        nativeWaitDurationMsTotal: AggregateCounterSchema,
        nativeWaitDurationMsMax: AggregateCounterSchema,
        modifierTimeouts: AggregateCounterSchema,
        failures: z
          .object({
            permissionDenied: AggregateCounterSchema,
            secureInput: AggregateCounterSchema,
            conflictingModifiers: AggregateCounterSchema,
            osRejected: AggregateCounterSchema,
            unavailable: AggregateCounterSchema,
            indeterminate: AggregateCounterSchema,
          })
          .strict(),
      })
      .strict(),
  })
  .strict()
  .superRefine((value, context) => {
    if (value.paste.nativeWaitDurationMsMax > value.paste.nativeWaitDurationMsTotal) {
      context.addIssue({
        code: 'custom',
        path: ['paste', 'nativeWaitDurationMsMax'],
        message: 'Maximum native modifier wait cannot exceed total wait duration',
      });
    }
  });

export const HelperTerminalObservabilityRecordSchema = z
  .object({
    event: z.literal('helper.runtime.terminal'),
    outcome: z.enum(['shutdown', 'failure']),
    observability: HelperRuntimeObservabilitySchema,
  })
  .strict();

const HelperDiagnosticDimensionsSchema = z
  .object({
    category: z.enum([
      'connect',
      'disconnected',
      'uncertain',
      'protocol',
      'acquire_rejected',
      'rejected',
      'sequence_exhausted',
      'transport',
    ]),
    operation: z.enum([
      'capture.replace_configuration',
      'capture.set_enabled',
      'command.sequence',
      'connect.reconcile',
      'established_operation',
      'front_app.get',
      'front_app.metadata.get',
      'front_app.metadata_get',
      'health.get',
      'lease.acquire',
      'lease.release',
      'lease.renew',
      'observability.get',
      'paste.await_commit',
      'paste.inject',
      'permissions.get',
      'runtime.rollback',
      'service',
      'service.poll',
      'session.reconcile_off',
      'session.set_mode',
    ]),
    correlationStatus: z.enum([
      'none',
      'not_established',
      'pending',
      'matched',
      'matched_initial_response',
      'mismatched',
      'unexpected_response',
      'unknown',
    ]),
    healthRefresh: z.enum(['not_attempted', 'succeeded', 'failed']),
    transportStatus: z.enum(['open', 'eof', 'closed', 'error', 'backpressured', 'unknown']),
    ownerProcessState: z.enum(['running', 'exited', 'unknown']),
  })
  .strict();
export const HelperDiagnosticAckParamsSchema = z
  .object({
    journalId: HelperDiagnosticIdentitySchema,
    journalNonce: HelperDiagnosticIdentitySchema,
    dimensions: HelperDiagnosticDimensionsSchema,
    count: z.string().regex(/^(?:0|[1-9][0-9]{0,38})$/u),
  })
  .strict();

File: helper/owner-protocol/src/schema/observability.rs
//! Privacy-safe bounded aggregate response schemas.
use super::*;

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct AuthFailures {
    pub cross_user: Counter,
    pub wrong_session: Counter,
    pub code_identity: Counter,
    pub mac: Counter,
    pub protocol: Counter,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct OwnerCounters {
    pub starts: Counter,
    pub clean_exits: Counter,
    pub abnormal_exits: Counter,
    pub singleton_collisions: Counter,
    pub auth_attempts: Counter,
    pub auth_failures: AuthFailures,
    pub lease_acquired: Counter,
    pub lease_renewed: Counter,
    pub lease_expired: Counter,
    pub lease_disconnected: Counter,
    pub lease_released_neutral: Counter,
    pub lease_released_draining: Counter,
    pub drain_duration_ms_total: Counter,
    pub drain_duration_ms_max: Counter,
    pub maintenance_postponed: Counter,
    pub handoff_succeeded: Counter,
    pub handoff_failed: Counter,
    pub degraded: Counter,
    pub hook_recoveries: Counter,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct CancellationReasons {
    pub invalid_continuation: Counter,
    pub modifier_changed: Counter,
    pub alt_gr: Counter,
    pub journal_overflow: Counter,
    pub configuration_replaced: Counter,
    pub revision_mismatch: Counter,
    pub gate_closed: Counter,
    pub shutdown: Counter,
    pub helper_disconnected: Counter,
    pub secure_desktop: Counter,
    pub timeout: Counter,
    pub activation_delivery_failed: Counter,
    pub neutralization_failed: Counter,
    pub replay_failed: Counter,
    pub effect_protocol_violation: Counter,
    pub physical_state_mismatch: Counter,
    pub target_changed: Counter,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct TransactionCounters {
    pub started: Counter,
    pub committed: Counter,
    pub replayed: Counter,
    pub cancelled: Counter,
    pub journal_high_water: Counter,
    pub cancellation_reasons: CancellationReasons,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct EffectCounters {
    pub attempted: Counter,
    pub succeeded: Counter,
    pub partial: Counter,
    pub failed: Counter,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct RegisteredInputCounters {
    pub hook_installed: Counter,
    pub pump_alive: Counter,
    pub hc_action_callbacks: Counter,
    pub physical_callbacks: Counter,
    pub physical_callbacks_filtered: Counter,
    pub registered_candidate_callbacks: Counter,
    pub registered_match_callbacks: Counter,
    pub registered_release_callbacks: Counter,
    pub callback_channel_accepted: Counter,
    pub callback_channel_rejected: Counter,
    pub adapter_dequeued: Counter,
    pub owner_admitted: Counter,
    pub owner_flushed: Counter,
    pub owner_rejected: Counter,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct NativePasteCounters {
    pub target_validation_fallbacks: Counter,
    pub modifier_wait_duration_ms_total: Counter,
    pub modifier_wait_duration_ms_max: Counter,
    pub modifier_timeouts: Counter,
    pub shutdown_ownership_deadlines: Counter,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ObservabilityResult {
    pub owner: OwnerCounters,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        rename = "registeredInput"
    )]
    pub registered_input: Option<RegisteredInputCounters>,
    pub transactions: TransactionCounters,
    pub replay: EffectCounters,
    pub dummy: EffectCounters,
    #[serde(rename = "nativePaste")]
    pub native_paste: NativePasteCounters,
}
